### PR TITLE
Sensing skystone and teleop changes

### PR DESCRIPTION
**Pull request by kevuyt**
Before issuing a pull request, please see the contributing page.

### CODE REVIEW

File: Masquerade4997CV/src/main/java/MasqCV/detectors/MasqCVDetector.java [hunk 1]
```diff
@@ -111,4 +112,14 @@ public void onViewportTapped()
...
     }
+
+    public Mat cropMat(Mat input, Point tl, Point br)  {
+        if (!(tl == null || br == null || tl.x >= input.width() || tl.y >= input.height() || tl.x < 0 || tl.y < 0 || br.x > input.width() || br.y > input.height() || br.x <= 0 || br.y <= 0)) {
+            Imgproc.rectangle(input,new Point(tl.x, tl.y),new Point(br.x, br.y),new Scalar(0), -1);
+        }
+        else {
+            DashBoard.getDash().create("Cropping failed due to invalid cropping margins");
+        }
+        return input;
```
Comment (archishou) [inline Masquerade4997CV/src/main/java/MasqCV/detectors/MasqCVDetector.java]: Where'd that come from?

File: Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetector.java [hunk 2]
```diff
@@ -20,20 +36,61 @@ public void setInput(Mat input) {
...
             Mat mat1 = input.submat(new Rect(0, 0, width/3, input.height()));
-            /*Mat mat2 = input.submat(new Rect(width/3, 0, width/3, input.height()));
+            Mat mat2 = input.submat(new Rect(width/3, 0, width/3, input.height()));
             Mat mat3 = input.submat(new Rect( 2 * (width/3), 0, width/3, input.height()));
             List<Mat> src = Arrays.asList(mat1, mat2, mat3);
-            Core.hconcat(src, input);*/
+            Core.hconcat(src, input);
             input = mat1;
-            DashBoard.getDash().create("Hello");
+            DashBoard.getDash().create("Hello");*/
```
Comment (archishou) [inline Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetector.java]: Delete

File: Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetectorv2.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+import org.opencv.core.Scalar;
+import org.opencv.imgproc.Imgproc;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.LinkedList;
+
+import Library4997.MasqWrappers.DashBoard;
+
+public class SkystoneDetectorv2 extends DogeCVDetector {
```
Comment (archishou) [inline Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetectorv2.java]: Delete this and move into SkystoneDetector.

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildStoneAuto.java [hunk 4]
```diff
@@ -20,10 +20,8 @@ public void runLinearOpMode() throws InterruptedException{
         robot.init(hardwareMap);
         robot.initializeAutonomous();
 
-        //while(!opModeIsActive()) {
-            dash.create("Hello");
-            dash.update();
-        //}
+        dash.create("Hello");
+        dash.update();
```
Comment (archishou) [inline Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildStoneAuto.java]: Nope. Use the while !opModeIsActive

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildWallAuto.java [hunk 5]
```diff
@@ -20,28 +20,28 @@ public void runLinearOpMode() throws InterruptedException{
         robot.init(hardwareMap);
         robot.initializeAutonomous();
 
-        while(!opModeIsActive()) {
-            dash.create("Hello");
-            dash.update();
-        }
+        dash.create("Hello");
```
Comment (archishou) [inline Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildWallAuto.java]: Nope

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Robot/MarkOne.java [hunk 6]
```diff
@@ -23,7 +23,7 @@
  */
 public class MarkOne extends MasqRobot {
 
-    public MasqServo blockGrabber, blockRotater, blockPusher, sideGrabber;
+    public MasqServo blockGrabber, blockRotater, blockPusher, capper;
```
Comment (archishou) [inline Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Robot/MarkOne.java]: lmao... capper

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Test/SkystoneDetectionTest.java [hunk 7]
```diff
@@ -29,13 +30,12 @@ public void runLinearOpMode() throws InterruptedException {
...
+            /*skystonePosition = robot.detector.skystoneDetector.getSkystonePosition();
             if (skystonePosition == null) robot.runStoneLeft(hardwareMap);
-            else if (skystonePosition == DogeDetector.SkystonePosition.LEFT) robot.runStoneLeft(hardwareMap);
-            else if (skystonePosition == DogeDetector.SkystonePosition.RIGHT) robot.runStoneRight(hardwareMap);
-            else if (skystonePosition == DogeDetector.SkystonePosition.MIDDLE) robot.runStoneMiddle(hardwareMap);
-            else robot.runStoneLeft(hardwareMap);
+            else if (skystonePosition == SkystoneDetector.SkystonePosition.LEFT) robot.runStoneLeft(hardwareMap);
+            else if (skystonePosition == SkystoneDetector.SkystonePosition.RIGHT) robot.runStoneRight(hardwareMap);
+            else if (skystonePosition == SkystoneDetector.SkystonePosition.MIDDLE) robot.runStoneMiddle(hardwareMap);
+            else robot.runStoneLeft(hardwareMap);*/
```
Comment (archishou) [inline Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Test/SkystoneDetectionTest.java]: ? Why is this commented out. Isn't that useful?

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Test/TensorFlowTeleop.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                    if (recognition.getLabel().equals("Skystone")) {
+                        if (recognition.getLeft() < 250) {
+                            position = SkystonePosition.LEFT;
+                        }
+                        else if (recognition.getLeft() > 250) {
+                            position = SkystonePosition.MIDDLE;
+                        }
+                    }
+                }
+            }
```
Comment (archishou) [inline Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Test/TensorFlowTeleop.java]: Wow.

File: Masquerade4997Library/src/main/java/Library4997/MasqDriveTrains/MasqMechanumDriveTrain.java [hunk 9]
```diff
@@ -49,6 +49,29 @@ public void setVelocityMECH(double angle, double speed, double targetHeading) {
...
+            leftFront /= max;
+            leftBack /= max;
+            rightFront /= max;
+            rightBack /= max;
+        }
+        leftDrive.motor1.setVelocity(leftFront);
+        leftDrive.motor2.setVelocity(leftBack);
+        rightDrive.motor1.setVelocity(rightFront);
+        rightDrive.motor2.setVelocity(rightBack);
+    }
```
Comment (archishou) [inline Masquerade4997Library/src/main/java/Library4997/MasqDriveTrains/MasqMechanumDriveTrain.java]: Why was this added? What changed?

File: Masquerade4997Library/src/main/java/Library4997/MasqRobot.java [hunk 10]
```diff
@@ -32,38 +32,45 @@
     public abstract void mapHardware(HardwareMap hardwareMap);
     public abstract void init(HardwareMap hardwareMap);
     public MasqMechanumDriveTrain driveTrain;
-    protected MasqPositionTracker tracker;
+    public MasqPositionTracker tracker;
     public DashBoard dash;
-    private double speedMultiplier = 1.414;
-    private double turnMultiplier = 1.414;
+    public double speedMultiplier = 1.414;
+    public double turnMultiplier = 1.414;
```
Comment (archishou) [inline Masquerade4997Library/src/main/java/Library4997/MasqRobot.java]: Move this to MasqUtils

File: Masquerade4997Library/src/main/java/Library4997/MasqRobot.java [hunk 11]
```diff
@@ -32,38 +32,45 @@
...
+            else if (timeoutTimer.elapsedTime(timeOut, MasqClock.Resolution.SECONDS)) bool = false;
+            else bool = Math.abs(angularError) > 0.01 || clicksRemaining/targetClicks > 0.05;
+        } while (bool);
         driveTrain.stopDriving();
         sleep(MasqUtils.DEFAULT_SLEEP_TIME);
     }
-    public void strafe(double distance, Strafe angle, double timeout, double speed) {
-        strafe(distance, angle.value, timeout, speed);
+    public void strafe(double distance, Strafe angle, double timeOut, double speed) {
+        strafe(distance, angle.value, timeOut, speed);
```
Comment (archishou) [inline Masquerade4997Library/src/main/java/Library4997/MasqRobot.java]: timeout is one word. keep it as timeout not timeOut

File: Masquerade4997Library/src/main/java/Library4997/MasqRobot.java [hunk 12]
```diff
@@ -92,8 +100,14 @@ public void drive(double distance, double speed, Direction direction, double tim
             dash.create("LEFT POWER: ", leftPower);
             dash.create("RIGHT POWER: ", rightPower);
             dash.create("ERROR: ", clicksRemaining);
+            dash.create("HEADING: ", tracker.getHeading());
             dash.update();
-        } while (opModeIsActive() && !timeoutTimer.elapsedTime(timeOut, MasqClock.Resolution.SECONDS) && (clicksRemaining / targetClicks) > 0.05);
+            if (!opModeIsActive()) bool = false;
+            else if (timeoutTimer.elapsedTime(timeOut, MasqClock.Resolution.SECONDS)) bool = false;
+            else bool = Math.abs(angularError) > 0.01 || clicksRemaining/targetClicks > 0.05;
+            //Only if the robot has reached its destination/Op Mode has stopped/Timed Out does it set this boolean
```
Comment (archishou) [inline Masquerade4997Library/src/main/java/Library4997/MasqRobot.java]: clean this code.

File: Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetector.java (hunk 2 above)
Comment (kevuyt) [reply Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetector.java]: I made it

Comment (kevuyt) [reply Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetector.java]: The whole file?

File: Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetectorv2.java (hunk 3 above)
Comment (kevuyt) [reply Masquerade4997CV/src/main/java/MasqCV/detectors/skystone/SkystoneDetectorv2.java]: Okay

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildStoneAuto.java (hunk 4 above)
Comment (kevuyt) [reply Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildStoneAuto.java]: It doesn’t change though

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Test/SkystoneDetectionTest.java (hunk 7 above)
Comment (kevuyt) [reply Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Test/SkystoneDetectionTest.java]: It gave some error and I’m not using it now so when I use it I will I comment it

File: Masquerade4997Library/src/main/java/Library4997/MasqRobot.java (hunk 10 above)
Comment (kevuyt) [reply Masquerade4997Library/src/main/java/Library4997/MasqRobot.java]: Okay

File: Masquerade4997CV/src/main/java/MasqCV/detectors/MasqCVDetector.java (hunk 1 above)
Comment (kevuyt) [reply Masquerade4997CV/src/main/java/MasqCV/detectors/MasqCVDetector.java]: I edited one of the screenshots you sent me

File: Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildStoneAuto.java (hunk 4 above)
Comment (archishou) [reply Masquerade4997Code/src/main/java/org/firstinspires/ftc/teamcode/Robots/MarkOne/Autonomous/Blue/BlueBuildStoneAuto.java]: While !opModeIsActive is standard. Use it
